### PR TITLE
feat(cli): polite deep merge for settings.json and support JSONC

### LINKED ISSUE #1799: [Feature]: vscode  settings.json support upgrade, don't use replace the content.
### Problem Statement

Image
when I upgrade the spec-kit new version, and try init by upgrade doc.
my viscose's setting will replace by script.

### Proposed Solution

so can we have a upgrade for this content? find the key can upgrade or something?

### Alternatives Considered

_No response_

### Component

Specify CLI (initialization, commands)

### AI Agent (if applicable)

Gemini CLI

### Use Cases

_No response_

### Acceptance Criteria

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by RbBtSn0w**
### Description

This PR addresses issue #1799, where the Specify CLI would overwrite the user's existing `.vscode/settings.json` file during initialization or upgrades, leading to data loss of personal preferences.

### Key Changes & Architectural Decisions

*   **JSONC Parsing Refactor (The `json5` library):** 
    Initially, a custom state-machine tokenizer was implemented to handle VS Code's JSONC format (comments and trailing commas). However, given the complexity of edge cases (e.g., trailing commas immediately followed by block comments before a closing brace) and to future-proof the parser against other JSON5/JSONC quirks, the custom parser was entirely removed. 
    Instead, **the robust `json5` library has been integrated**. It natively, effortlessly, and safely parses files with single/multi-line comments, trailing commas, and BOMs without the maintenance overhead of a custom tokenizer.
*   **Polite Deep Merge:** Implemented `deep_merge_polite` which strictly ensures that:
    *   New recommended settings from templates are added.
    *   **Existing user settings are preserved** and are never overwritten by template defaults.
    *   Nested dictionaries are merged recursively to maintain complex structures.
*   **Non-Destructive Fallbacks (Zero Data Loss Guarantee):** 
    Added extreme defensive programming to the merge pipeline. If the user's existing `settings.json` is completely unparseable (e.g., severe syntax error) or is not a JSON object (dictionary), the `merge_json_files` function immediately returns `None`. The CLI catches this and gracefully skips the file writing step entirely, guaranteeing that the user's original (even if invalid) configuration is never wiped by an empty dictionary or overridden by the template.
*   **Unit Tests:** Added comprehensive tests in `tests/test_merge.py` that specifically target the deep-merge preservation rules, type mismatch handling, and the secure fallback behavior.

### Verification

*   Verified that `.vscode/settings.json` containing comments and trailing commas can be parsed cleanly via `json5`.
*   Verified that user-defined keys (e.g., `editor.fontSize`) remain strictly intact.
*   Verified that unparseable user files do not result in a destructive overwrite.
*   Verified all tests pass with `uv run pytest tests/test_merge.py`.

Fixes #1799

### CODE REVIEW

**Comment by RbBtSn0w** [comment]
I've addressed the feedback regarding JSONC parsing:
- Implemented a state-machine tokenizer for `strip_json_comments` that correctly handles string literals (e.g., URLs).
- Added support for trailing commas in objects and arrays.
- Fixed indentation issues in `src/specify_cli/__init__.py`.
- Added new test cases in `tests/test_merge.py` to cover these scenarios.

**Comment by RbBtSn0w** [comment]
Expanded test coverage across multiple dimensions:
- **String literals**: Verified URL-like strings and escaped characters are not corrupted.
- **Trailing commas**: Verified support for trailing commas in nested objects and arrays.
- **BOM handling**: Added support for UTF-8 Byte Order Mark (BOM).
- **Merge conflicts**: Verified polite merging preserves existing settings when types mismatch (e.g., string vs object).
- **Realistic scenarios**: Simulated complex `settings.json` file merging.
All 9 test cases in `tests/test_merge.py` are passing.

**Comment by RbBtSn0w** [comment]
Addressing final review comments:
- **Refined Backslash Handling**: `strip_json_comments` now determines whether a quote is escaped by correctly counting consecutive backslashes (odd = escaped, even = terminator). Added test cases for `even_backslashes` and `odd_backslashes`.
- **Improved Validation**: `merge_json_files` now explicitly validates that `new_content` is a dictionary and returns an empty dictionary with a warning if it is not.
- **Cleanup**: Removed the unused `pytest` import in `tests/test_merge.py`.
- **Refined Test Cases**: Updated tests to reflect the new validation logic and cover complex escape sequences.

**Comment by RbBtSn0w** [comment]
Final updates to address all Copilot and Maintainer concerns:
- **String-Safe Trailing Commas**: Integrated trailing comma removal into the state-machine loop of `strip_json_comments`. This ensures commas inside strings (e.g. `"val, ]"`) are strictly preserved.
- **Secure Fallback Mechanism**: Updated `merge_json_files` to return the existing configuration instead of an empty dictionary when the template (`new_content`) is invalid. This prevents accidental data loss.
- **Typo Fix**: Corrected "Correctlly" to "Correctly".
- **Robustness**: Added a specific test case `test_strip_json_comments_safe_commas_in_strings` to verify the state-machine's precision.
- **Indentation**: Normalized all indentations to 4-space multiples.

**Review by mnriem — CHANGES_REQUESTED** [review]
Please address Copilot feedback. If not applicable, please explain why

**Comment by RbBtSn0w** [comment]
All Copilot feedback has been addressed:
- Removed unused imports (`re` in `__init__.py`, and `Path`/`pytest` in `test_merge.py`).
- Fixed the fallback behavior so that it acts securely as a non-destructive operation: if the template or the existing JSON is invalid/non-dict, it safely returns `None` and the merge is skipped, ensuring user's data is strictly preserved and never overwritten with `{}`.
- Added edge-case test coverage for an unparseable existing `settings.json`.

All tests pass. Could you please take another look?

File: tests/test_merge.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                         comment */
+        "c": 3 // Another one
+    }
+    """
+    stripped = strip_json_comments(jsonc)
+    parsed = json.loads(stripped)
+    assert parsed["a"] == 1
+    assert parsed["b"] == "value"
+    assert parsed["c"] == 3
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py:22]: Addressed in previous commits. The test suite (tests/test_merge.py) now includes comprehensive test cases like test_strip_json_comments_edge_cases,
  test_strip_json_comments_trailing_commas_advanced, and test_strip_json_comments_safe_commas_in_strings which specifically cover string literals with URLs (//), paths,
  nested comments (/* */), and edge-case trailing commas.

File: src/specify_cli/__init__.py [hunk 2]
```diff
@@ -696,24 +711,42 @@ def merge_json_files(existing_path: Path, new_content: dict, verbose: bool = Fal
...
+            
+        if not isinstance(existing_content, dict):
+             if verbose:
+                 console.print(f"[yellow]Warning: Existing JSON in {existing_path.name} is not an object, using template instead[/yellow]")
+             return new_content
+             
+    except (FileNotFoundError, json.JSONDecodeError) as e:
         # If file doesn't exist or is invalid, just use new content
+        if verbose and not isinstance(e, FileNotFoundError):
+             console.print(f"[yellow]Warning: Could not parse existing JSON in {existing_path.name} ({e}), using template instead[/yellow]")
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. The indentation issues have been fixed in a previous commit. All if verbose: and console.print(...) blocks are now correctly aligned with standard 4-space
  indentations, and will pass linting checks.

[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. Just like the previous comment, this indentation issue has been fixed and normalized to 4-space multiples in a prior commit. Everything is correctly
  aligned now and ready for linting.

File: src/specify_cli/__init__.py [hunk 3]
```diff
@@ -677,14 +680,26 @@ def log(message, color="green"):
...
+    
+    This is a basic implementation to handle VSCode settings.json files.
+    """
+    import re
+    # Remove single-line comments
+    content = re.sub(r'//.*$', '', content, flags=re.MULTILINE)
+    # Remove multi-line comments
+    content = re.sub(r'/\*.*?\*/', '', content, flags=re.DOTALL)
+    return content
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed in a previous commit. The regex-based approach has been completely replaced with a custom state-machine tokenizer. It now safely tracks string boundaries,
  correctly handles escape sequences (even with trailing backslashes), preserves URLs (//) inside strings, and securely removes single/multi-line comments and trailing
  commas outside of string literals.

File: CHANGELOG.md [hunk 4]
```diff
@@ -35,6 +35,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- feat(cli): support for JSONC (JSON with comments) in VSCode settings.json merging
+- feat(cli): polite deep merge for JSON files that preserves existing user settings
+- feat(cli): basic validation for JSON object structures during initialization
+- feat(cli): strip_json_comments utility for handling VSCode configuration files
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply CHANGELOG.md]: Addressed in a previous commit. The implementation has been expanded beyond just comment stripping. The state-machine tokenizer now accurately detects and removes trailing commas from arrays and objects (while preserving them if they occur inside string literals). This makes it fully compatible with standard VS Code JSONC files. Test cases covering advanced trailing comma scenarios have also been added.

File: tests/test_merge.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        "array": [1, 2, 3, ],
+        "obj": {
+            "a": 1,
+        },
+    }
+    """
+    stripped = strip_json_comments(jsonc)
+    parsed = json.loads(stripped)
+    assert parsed["array"] == [1, 2, 3]
+    assert parsed["obj"]["a"] == 1
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: Addressed in a previous commit. The regex-based approach was replaced with a robust state-machine tokenizer that accurately counts consecutive backslashes (odd=escaped, even=terminator) and safely handles trailing commas only when strictly outside of string literals.

File: tests/test_merge.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
+import json
+from pathlib import Path
+import pytest
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: Addressed. The unused imports have been removed to keep the codebase clean and avoid triggering linting checks.

File: src/specify_cli/__init__.py [hunk 7]
```diff
@@ -677,14 +680,75 @@ def log(message, color="green"):
...
+    while i < length:
+        char = content[i]
+        
+        # Handle string literals
+        if char in ('"', "'") and (i == 0 or content[i-1] != '\\'):
+            if not in_string:
+                in_string = True
+                string_char = char
+            elif char == string_char:
+                in_string = False
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed in a previous commit. The regex-based approach was replaced with a robust state-machine tokenizer that accurately counts consecutive backslashes (odd=escaped, even=terminator) and safely handles trailing commas only when strictly outside of string literals.

File: src/specify_cli/__init__.py [hunk 8]
```diff
@@ -677,14 +680,75 @@ def log(message, color="green"):
...
+            
+        result.append(char)
+        i += 1
+        
+    clean_content = "".join(result)
+    
+    # Handle trailing commas: find a comma followed by closing brace/bracket
+    # This regex handles cases like [1, 2, ] or {"a": 1, }
+    clean_content = re.sub(r',\s*([\]}])', r'\1', clean_content)
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed in a previous commit. The regex-based approach was replaced with a robust state-machine tokenizer that accurately counts consecutive backslashes (odd=escaped, even=terminator) and safely handles trailing commas only when strictly outside of string literals.

File: tests/test_merge.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    merged = merge_json_files(existing_file, new_settings)
+    assert merged == {"a": 1, "b": 2}
+
+def test_merge_json_files_not_a_dictionary_template(tmp_path):
+    """If for some reason new_content is not a dict, return empty dict as safety."""
+    existing_file = tmp_path / "ok.json"
+    existing_file.write_text('{"a": 1}')
+    
+    # In the updated implementation, if new_content is not a dict, we return {}
+    assert merge_json_files(existing_file, ["not", "a", "dict"]) == {}
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: Addressed. The fallback logic has been completely overhauled. If  is invalid, or if the existing settings file is unparseable or not a dictionary,  now strictly returns . The  caller catches this  and safely skips the file write entirely, guaranteeing zero data loss.

File: src/specify_cli/__init__.py [hunk 10]
```diff
@@ -677,14 +680,96 @@ def log(message, color="green"):
         log(f"Warning: Could not merge, copying instead: {e}", "yellow")
         shutil.copy2(sub_item, dest_file)
 
+def strip_json_comments(content: str) -> str:
+    """Remove // and /* */ comments from a JSON string, respecting string literals.
+    
+    This implementation handles VSCode settings.json files by:
+    1. Skipping comments inside "string literals"
+    2. Correctlly handling escaped quotes by counting backslashes
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. The spelling error was fixed in a previous commit.

File: src/specify_cli/__init__.py [hunk 11]
```diff
@@ -677,14 +680,96 @@ def log(message, color="green"):
...
+    # Handle trailing commas: find a comma followed by closing brace/bracket
+    # This is now safer as we've already stripped comments and identified strings
+    # However, to be 100% safe against commas in strings, we should have handled this 
+    # during the character loop. Let's refine the regex to be very specific to 
+    # JSON structure (comma followed by whitespace and closing char).
+    # Since we've already joined the result, we can apply a final cleanup.
+    # To truly avoid commas in strings, we'd need to mark string ranges.
+    # For now, this regex is standard for basic JSONC to JSON conversion.
+    clean_content = re.sub(r',\s*([\]}])', r'\1', clean_content)
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. The fallback logic has been completely overhauled. If  is invalid, or if the existing settings file is unparseable or not a dictionary,  now strictly returns . The  caller catches this  and safely skips the file write entirely, guaranteeing zero data loss.

File: src/specify_cli/__init__.py [hunk 12]
```diff
@@ -694,26 +779,49 @@ def merge_json_files(existing_path: Path, new_content: dict, verbose: bool = Fal
     Returns:
         Merged JSON content as dict
     """
+    if not isinstance(new_content, dict):
+        if verbose:
+            console.print(f"[yellow]Warning: Template content for {existing_path.name} is not a dictionary. Skipping merge.[/yellow]")
+        return {}
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py:765]: Addressed. The fallback logic has been completely overhauled. If  is invalid, or if the existing settings file is unparseable or not a dictionary,  now strictly returns . The  caller catches this  and safely skips the file write entirely, guaranteeing zero data loss.

File: src/specify_cli/__init__.py [hunk 13]
```diff
@@ -694,26 +779,49 @@ def merge_json_files(existing_path: Path, new_content: dict, verbose: bool = Fal
...
+            with open(existing_path, 'r', encoding='utf-8') as f:
+                raw_content = f.read().lstrip('\ufeff')
+                # Handle comments (JSONC)
+                clean_content = strip_json_comments(raw_content)
+                existing_content = json.loads(clean_content)
+                
+            if not isinstance(existing_content, dict):
+                if verbose:
+                    console.print(f"[yellow]Warning: Existing JSON in {existing_path.name} is not an object, using template instead[/yellow]")
+                return new_content
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. The fallback logic has been completely overhauled. If  is invalid, or if the existing settings file is unparseable or not a dictionary,  now strictly returns . The  caller catches this  and safely skips the file write entirely, guaranteeing zero data loss.

File: src/specify_cli/__init__.py [hunk 14]
```diff
@@ -677,14 +680,102 @@ def log(message, color="green"):
...
+def strip_json_comments(content: str) -> str:
+    """Remove // and /* */ comments from a JSON string, respecting string literals.
+    
+    This implementation handles VSCode settings.json files by:
+    1. Skipping comments inside "string literals"
+    2. Correctly handling escaped quotes by counting backslashes
+    3. Removing single-line (//) and multi-line (/* */) comments
+    4. Handling trailing commas safely (only outside strings/comments)
+    """
+    import re
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. The unused imports have been removed to keep the codebase clean and avoid triggering linting checks.

File: tests/test_merge.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
+import json
+from pathlib import Path
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: Addressed. The unused imports have been removed to keep the codebase clean and avoid triggering linting checks.

File: tests/test_merge.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+    merged = merge_json_files(existing_file, new_settings)
+    assert merged == {"a": 1, "b": 2}
+
+def test_merge_json_files_not_a_dictionary_template(tmp_path):
+    """If for some reason new_content is not a dict, PRESERVE existing settings instead of wiping."""
+    existing_file = tmp_path / "ok.json"
+    existing_file.write_text('{"a": 1}')
+    
+    # Secure fallback: return existing settings to avoid clobbering
+    assert merge_json_files(existing_file, ["not", "a", "dict"]) == {"a": 1}
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: Addressed. I've updated the fallback test to assert that  is returned to prevent clobbering. I also added  to specifically cover the scenario where the existing file is syntactically invalid, ensuring the merge is safely bypassed.

File: src/specify_cli/__init__.py [hunk 17]
```diff
@@ -696,24 +760,42 @@ def merge_json_files(existing_path: Path, new_content: dict, verbose: bool = Fal
...
+             
+    except (FileNotFoundError, json.JSONDecodeError) as e:
         # If file doesn't exist or is invalid, just use new content
+        if verbose and not isinstance(e, FileNotFoundError):
+            console.print(f"[yellow]Warning: Could not parse existing JSON in {existing_path.name} ({e}), using template instead[/yellow]")
         return new_content
 
-    def deep_merge(base: dict, update: dict) -> dict:
-        """Recursively merge update dict into base dict."""
+    if not isinstance(new_content, dict):
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py:761]: Addressed. The function signature has been updated to return `Optional[dict]`. When `new_content` is not a dict, it now logs a visible warning and explicitly returns `None` instead of falling back to `existing_content`. This signals a failure to the caller (`handle_vscode_settings`), which then safely aborts the merge and prevents any data loss.

File: src/specify_cli/__init__.py [hunk 18]
```diff
@@ -694,26 +785,54 @@ def merge_json_files(existing_path: Path, new_content: dict, verbose: bool = Fal
...
+            with open(existing_path, 'r', encoding='utf-8') as f:
+                raw_content = f.read().lstrip('\ufeff')
+                # Handle comments (JSONC)
+                clean_content = strip_json_comments(raw_content)
+                existing_content = json.loads(clean_content)
+        except (FileNotFoundError, json.JSONDecodeError) as e:
+            if verbose and not isinstance(e, FileNotFoundError):
+                console.print(f"[yellow]Warning: Could not parse existing JSON in {existing_path.name} ({e}).[/yellow]")
+            # fallback to empty dict for merging
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py:759]: Addressed. The fallback logic has been updated to strictly act as a non-destructive operation. If the existing file is unparseable (e.g. `JSONDecodeError`), `merge_json_files()` now explicitly returns `None`. The caller (`handle_vscode_settings`) checks for this and gracefully skips writing to the file entirely, effectively preserving the user's original unparseable file without any data loss. I also added `test_merge_json_files_unparseable_existing` to guarantee this behavior.

**Review by RbBtSn0w — COMMENTED** [review]
I have addressed the final edge case regarding trailing commas followed by comments. The `strip_json_comments` state-machine lookahead has been updated to safely skip over `//` and `/* */` blocks when searching for the closing bracket/brace. Associated tests have been added.

File: tests/test_merge.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+            "tags": ["a", "b", ],
+        },
+    }
+    """
+    stripped = strip_json_comments(jsonc)
+    parsed = json.loads(stripped)
+    assert len(parsed["list"]) == 2
+    assert parsed["list"][0]["id"] == 1
+    assert parsed["metadata"]["tags"] == ["a", "b"]
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: Addressed. I've added the `test_strip_json_comments_trailing_comma_before_comment` test case which specifically covers both the inline (`/* */`) and single-line (`//`) comment scenarios immediately following a trailing comma before a closing bracket/brace. The `strip_json_comments()` implementation was updated to successfully pass this test.

**Review by RbBtSn0w — COMMENTED** [review]
I have completely refactored the JSON parsing approach. To eliminate all edge cases and future-proof the JSONC compatibility (including complex comment patterns and trailing commas), I removed the custom state-machine implementation and integrated the `json5` library as a dependency. It natively and safely handles the VS Code JSONC format. All merge strategies and non-destructive fallbacks remain intact.

**Review by RbBtSn0w — COMMENTED** [review]
### Refactoring Update: Replacing Custom Tokenizer with `json5`

During the review process, we identified several edge cases in the custom state-machine used to parse JSONC (specifically around Lookaheads for trailing commas followed by block comments). 

To eliminate these edge cases entirely, reduce code complexity, and future-proof the CLI against any other VS Code JSONC/JSON5 quirks, **I have completely removed the custom tokenizer and integrated the `json5` library.** 

**Why `json5`?**
1. **Bulletproof Parsing:** It natively handles single/multi-line comments, trailing commas, BOMs, and even unquoted keys if they ever occur.
2. **Reduced Maintenance:** We no longer need to maintain 100+ lines of state-machine logic and the associated parsing unit tests. 
3. **Simplicity:** The file read logic is now as simple as `json5.load(f)`.

The core business logic—the **Polite Deep Merge** and the **Non-destructive fallbacks** (returning `None` to safely skip overwriting bad files)—remains fully intact. The test suite has been updated to focus purely on the merge strategies rather than parsing mechanics.

All CI checks should now pass cleanly without any linting or edge-case failures.

**Review by RbBtSn0w — COMMENTED** [review]
Addressed. I've removed the unused `import json` from `tests/test_merge.py` to keep the test module perfectly clean and free of lint warnings.

File: src/specify_cli/__init__.py [hunk 20]
```diff
@@ -661,59 +662,93 @@ def log(message, color="green"):
...
     except Exception as e:
-        log(f"Warning: Could not merge, copying instead: {e}", "yellow")
-        shutil.copy2(sub_item, dest_file)
+        log(f"Warning: Could not merge settings: {e}", "yellow")
+        if not dest_file.exists():
+            shutil.copy2(sub_item, dest_file)
 
-def merge_json_files(existing_path: Path, new_content: dict, verbose: bool = False) -> dict:
+
+def merge_json_files(existing_path: Path, new_content: dict, verbose: bool = False) -> Optional[dict]:
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. I've updated the type hint for `new_content` from `dict` to `Any` in `merge_json_files()`. This accurately reflects that the function performs its own runtime validation and can safely handle non-dictionary inputs by returning `None`.

File: tests/test_merge.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        "editor.fontSize": 12,
+        "editor.formatOnSave": true,
+        "files.exclude": {
+            "**/.git": true,
+            "**/node_modules": true
+        },
+        "chat.promptFilesRecommendations": {
+            "existing.tool": true
+        } // User comment
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: I've updated the test_merge_vscode_realistic_scenario fixture in tests/test_merge.py to include both block comments
  (/* */) and multiple trailing comma cases. This ensures that the json5-based merge flow is explicitly tested against a broader
  range of real-world VS Code JSONC patterns and won't regress.

File: src/specify_cli/__init__.py [hunk 22]
```diff
@@ -661,59 +662,94 @@ def log(message, color="green"):
...
+            elif isinstance(result[key], dict) and isinstance(value, dict):
                 # Recursively merge nested dictionaries
-                result[key] = deep_merge(result[key], value)
+                result[key] = deep_merge_polite(result[key], value)
             else:
-                # Add new key or replace existing value
-                result[key] = value
+                # Key already exists and is not a dict, PRESERVE existing value
+                # This ensures user settings aren't overwritten by template defaults
+                pass
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py:788]: Addressed. I've added a dedicated test case test_merge_json_files_list_preservation to tests/test_merge.py. It explicitly
  verifies that existing list values are strictly preserved and are neither overwritten by the template's list nor merged with it.
  This confirms our "polite" merge policy and guards against regressions for non-dictionary collections.

File: src/specify_cli/__init__.py [hunk 23]
```diff
@@ -661,59 +662,94 @@ def log(message, color="green"):
...
+        try:
+            with open(existing_path, 'r', encoding='utf-8') as f:
+                # Handle comments (JSONC) natively with json5
+                # Note: json5 handles BOM automatically
+                existing_content = json5.load(f)
+        except Exception as e:
+            if verbose and not isinstance(e, FileNotFoundError):
+                console.print(f"[yellow]Warning: Could not parse existing JSON in {existing_path.name} ({e}).[/yellow]")
+            # Skip merge to preserve existing file if unparseable
+            if not isinstance(e, FileNotFoundError):
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Excellent observation. This race condition could indeed lead to missing template settings if the file were deleted exactly
  between the exists() and open() calls. I've updated merge_json_files() to catch FileNotFoundError (and OSError) inside the if
  exists: block and explicitly set exists = False. This ensures the function correctly proceeds to return new_content as intended
  for missing files.

File: src/specify_cli/__init__.py [hunk 24]
```diff
@@ -661,59 +663,96 @@ def log(message, color="green"):
...
         if dest_file.exists():
             merged = merge_json_files(dest_file, new_settings, verbose=verbose and not tracker)
-            with open(dest_file, 'w', encoding='utf-8') as f:
-                json.dump(merged, f, indent=4)
-                f.write('\n')
-            log("Merged:", "green")
+            if merged is not None:
+                with open(dest_file, 'w', encoding='utf-8') as f:
+                    json.dump(merged, f, indent=4)
+                    f.write('\n')
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. I've implemented an atomic write pattern in `handle_vscode_settings()` using `tempfile.mkstemp()` and `os.replace()`. This ensures that the `settings.json` file is never left in a truncated or corrupted state if the process crashes during the write operation.

File: src/specify_cli/__init__.py [hunk 25]
```diff
@@ -661,59 +663,96 @@ def log(message, color="green"):
...
+
+    if exists:
+        try:
+            with open(existing_path, 'r', encoding='utf-8') as f:
+                # Handle comments (JSONC) natively with json5
+                # Note: json5 handles BOM automatically
+                existing_content = json5.load(f)
+        except (FileNotFoundError, OSError):
+            # Handle race condition where file is deleted after exists() check
+            exists = False
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. I've refined the exception handling in `merge_json_files()` to strictly differentiate between `FileNotFoundError` (handled as a missing file race condition) and other errors (like `PermissionError` or parsing issues). For the latter, the function now returns `None` to ensure the existing file is preserved and left untouched.

File: CHANGELOG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply CHANGELOG.md]: Addressed. I've updated the changelog to remove the stale reference to `strip_json_comments` and consolidated the entries to accurately reflect the use of `json5` and the polite deep merge strategy.

File: tests/test_merge.py [hunk 26]
```diff
@@ -0,0 +1,122 @@
+from specify_cli import merge_json_files
+
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py:5]: Addressed. I've removed the unused `import json` from the test file to keep it clean.

File: tests/test_merge.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply tests/test_merge.py]: Addressed. I've added block comments and trailing commas to the realistic VS Code test fixture to ensure full JSONC coverage.

File: src/specify_cli/__init__.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py:788]: Addressed. I've added `test_merge_json_files_list_preservation` to explicitly verify that existing lists are preserved.

File: src/specify_cli/__init__.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. I've updated the logic to handle the FileNotFoundError race condition by setting `exists = False` and proceeding.

File: src/specify_cli/__init__.py [hunk 27]
```diff
@@ -661,59 +663,94 @@ def log(message, color="green"):
...
         if dest_file.exists():
             merged = merge_json_files(dest_file, new_settings, verbose=verbose and not tracker)
-            with open(dest_file, 'w', encoding='utf-8') as f:
-                json.dump(merged, f, indent=4)
-                f.write('\n')
-            log("Merged:", "green")
+            if merged is not None:
+                with open(dest_file, 'w', encoding='utf-8') as f:
+                    json.dump(merged, f, indent=4)
+                    f.write('\n')
```
[the remark this answers is not in the document]
Comment (RbBtSn0w) [reply src/specify_cli/__init__.py]: Addressed. I've implemented change detection to skip the file rewrite if the merge results in no structural changes, thus preserving original formatting and comments in no-op cases.

**Comment by RbBtSn0w** [comment]
This PR is superseded by #1874 to provide a cleaner review history with the same final code changes. Please continue review there.

**Comment by RbBtSn0w** [comment]
Closing in favor of #1874 (same content, cleaner commit/review history).
